framework/sandstone.cpp

```diff
@@ -1042,6 +1047,87 @@ int num_packages() {
 }
 
 namespace {
+class BarrierDeviceSchedule : public DeviceSchedule
+{
+public:
+    void device_reschedule() override {
+        auto on_completion = [&]() noexcept {
+            std::unique_lock lock(groups_mutex);
+
+            int group_idx = thread_num / members_per_group;
+            GroupInfo *group = groups[group_idx];
+
+            // Rotate cpus vector so reschedule group members to a different cpu
+            std::rotate(group->next_cpu.begin(), group->next_cpu.begin() + 1, group->next_cpu.end());
+            lock.unlock();
+
+            // Reschedule group members
+            for (int i=0; i<members_per_group; i++) {
+                if(!pin_thread_to_logical_processor(LogicalProcessor(cpu_info[group->next_cpu[i]].cpu_number), group->pid[i]))
+                    log_debug("Failed to reschedule %d (%lld) to CPU %d", thread_num, pthread_self(), cpu_info[group->next_cpu[i]].cpu_number);
+            }
+        };
+
+        std::unique_lock lock(groups_mutex);
+
+        // Initialize groups on first run
+        if (groups.size() == 0) {
+            groups.reserve(num_cpus()/members_per_group);
+            for (int i=0; i<num_cpus(); i+=members_per_group) {
+                groups.push_back(new GroupInfo(members_per_group, on_completion));
```

Comment:
Any reason to use pointers here? If you can, make `groups` be a `vector<GroupInfo>` and here you can initialise each member with `emplace_back`:
```c++
    groups.emplace_back(members_per_group, on_completion);
```